src/v/iceberg/rest_catalog.cc

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "iceberg/rest_catalog.h"
+
+#include "iceberg/rest_client/catalog_client.h"
+#include "iceberg/table_requests.h"
+namespace iceberg {
+
+namespace {
+
+using http_status = enum boost::beast::http::status;
+using errc = enum iceberg::catalog::errc;
+using enum errc;
+
+struct http_error_mapping_visitor {
+    errc map_http_status(http_status status) const {
+        if (status == http_status::not_found) {
+            return not_found;
+        }
+
+        if (status == http_status::conflict) {
+            return already_exists;
+        }
+        return unexpected_state;
+    }
+
+    errc operator()(const rest_client::http_call_error& err) const {
+        return ss::visit(
+          err,
+          [this](http_status status) { return map_http_status(status); },
+          [](const ss::sstring&) { return unexpected_state; });
+    }
+
+    errc operator()(const rest_client::json_parse_error&) const {
+        return unexpected_state;
+    }
+
+    errc operator()(const rest_client::retries_exhausted&) const {
+        return timedout;
+    }
+
+    errc operator()(const http::url_build_error&) const {
+        return unexpected_state;
+    }
+};
+// Translates domain error to more general catalog::errc.
+errc map_error(
+  std::string_view context, const rest_client::domain_error& error) {
+    vlog(log.warn, "error returned when executing {} - {}", context, error);
+    return std::visit(http_error_mapping_visitor{}, error);
+}
+
+table_metadata get_metadata(load_table_result&& result) {
+    return std::move(result.metadata);
+}
+
+table_update::update copy(const table_update::update& update) {
+    return ss::visit(
+      update, [](const auto& u) { return table_update::update{u.copy()}; });
+}
+
+} // namespace
+
+rest_catalog::rest_catalog(
+  std::unique_ptr<rest_client::catalog_client> client,
```

Comment:
I'm thinking maybe this should be some kind of client factory. I recall at some point Abhijat was concerned about certain kinds of errors invalidating the client, in which case a pool could transparently shutdown the client and create a new one. A factory (or pool) would be useful in those cases. Will ask around for pointers

Reply:
i was considering that, maybe i can change that in a future PR ?